File: c/pangram/src/pangram.c
#include "pangram.h"
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

char *lower_case_letters(const char *string);

int flip_bit(int seen, char c);

const int TWENTY_SIX_BITS_FLIPPED = 67108863;

bool is_pangram(const char *sentence)
{
    if (!sentence)
        return false;

	int i, seen = 0;
	char *letters_only = lower_case_letters(sentence);

	for (i = 0; letters_only[i]; i++)
		seen = flip_bit(seen, letters_only[i]);

	free(letters_only);

	return seen == TWENTY_SIX_BITS_FLIPPED;
}

char *lower_case_letters(const char *string)
{
    int i, j = 0;
    char *letters_only = calloc(strlen(string), sizeof(char));

    for (i = 0; string[i]; i++)
        if (isalpha(string[i]))
            letters_only[j++] = tolower(string[i]);

    return letters_only;
}

int flip_bit(int seen, const char c)
{
    return seen | 1 << (c - 97);
}

File: go/pangram/pangram.go
// Package pangram contains a method checking for pangrams
package pangram

import (
	"strings"
	"unicode"
)

const seenAllTwentySixChars int = 0x3FFFFFF

// IsPangram checks if candidate contains all letters of the alphabet
func IsPangram(candidate string) bool {
	seen := 0

	for _, char := range strings.ToLower(candidate) {
		if unicode.IsLetter(char) {
			seen = flipBit(char, seen)
		}
	}

	return seen == seenAllTwentySixChars
}

func flipBit(char rune, seen int) int {
	return seen | 1<<uint(char-97)
}
